packages/core/src/connectivity/scp-cf/environment-accessor.ts

```diff
@@ -238,91 +237,125 @@ export function resolveService(service: string | Service): Service {
 export function extractClientCredentials(
   serviceCreds: ServiceCredentials
 ): ClientCredentials {
+  if (!serviceCreds.clientsecret) {
+    throw new Error(
+      `Cloud not extract client secret for clientId: ${serviceCreds.clientid}.`
+    );
+  }
   return {
     username: serviceCreds.clientid,
     password: serviceCreds.clientsecret
   };
 }
 
 function selectXsuaaInstance(token?: JwtPayload): XsuaaServiceCredentials {
-  const xsuaaInstances = getServiceList('xsuaa');
+  const xsuaaCredentials: XsuaaServiceCredentials[] = getServiceList(
+    'xsuaa'
+  ).map(service => service.credentials as XsuaaServiceCredentials);
 
-  if (!xsuaaInstances.length) {
+  if (!xsuaaCredentials.length) {
     throw Error(
       'No binding to an XSUAA service instance found. Please make sure to bind an instance of the XSUAA service to your application.'
     );
   }
 
-  const strategies = [matchingClientId, matchingAudience, takeFirstAndWarn];
-  const selected = applyStrategiesInOrder(strategies, xsuaaInstances, token);
+  if (token) {
+    return selectXsuaaInstanceWithJwt(xsuaaCredentials, token);
+  }
+  return selectXsuaaInstanceWithoutJwt(xsuaaCredentials);
+}
 
-  if (selected.length === 0) {
-    throw Error('No XSUAA instances are found from the given JWT.');
+function handleOneXsuuaInstance(
+  xsuaaCredentials: XsuaaServiceCredentials[]
+): XsuaaServiceCredentials {
```

Comment:
This function and the error it throws seems a bit unnecessary to me.
I assume you wanted to have this function to avoid duplicating the debug message and TS tells you that `xsuaaCredentials[0]` could be undefined?
Maybe this can be circumvented with this:
```suggestion
function handleOneXsuuaInstance(
  xsuaaCredentials: [XsuaaServiceCredentials]
): XsuaaServiceCredentials {
```

Reply:
yes this works. For more than one element you can use `[a,...a[]]` and with type guards you can select the cases.